src/NHibernate/Linq/Visitors/QueryModelVisitor.cs

```diff
@@ -128,16 +133,148 @@ static QueryModelVisitor()
 			ResultOperatorMap.Add<CastResultOperator, ProcessCast>();
 		}
 
-		private QueryModelVisitor(VisitorParameters visitorParameters, bool root, QueryModel queryModel)
+		private QueryModelVisitor(VisitorParameters visitorParameters, bool root, QueryModel queryModel,
+			NhLinqExpressionReturnType? rootReturnType)
 		{
 			VisitorParameters = visitorParameters;
 			Model = queryModel;
+			_rootReturnType = root ? rootReturnType : null;
 			_hqlTree = new IntermediateHqlTree(root);
 		}
 
 		private void Visit()
 		{
 			VisitQueryModel(Model);
+			AddAdditionalPostExecuteTransformer();
+		}
+
+		private void AddAdditionalPostExecuteTransformer()
+		{
+			if (_rootReturnType == NhLinqExpressionReturnType.Scalar && Model.ResultTypeOverride != null)
+			{
+				// NH-3850: handle polymorphic scalar results aggregation
+				switch ((NhExpressionType)Model.SelectClause.Selector.NodeType)
+				{
+					case NhExpressionType.Average:
+						// Polymorphic case complex to handle and not implemented. (HQL query must be reshaped for adding
+						// additional data to allow a meaningful overall average computation.)
+						// Leaving it untouched for allowing non polymorphic cases to work.
+						break;
+					case NhExpressionType.Max:
+						AddPostExecuteTransformerForResultAggregate(ReflectionCache.QueryableMethods.MaxDefinition);
+						break;
+					case NhExpressionType.Min:
+						AddPostExecuteTransformerForResultAggregate(ReflectionCache.QueryableMethods.MinDefinition);
+						break;
+					case NhExpressionType.Count:
+						AddPostExecuteTransformerForCount();
+						break;
+					case NhExpressionType.Sum:
+						AddPostExecuteTransformerForSum();
+						break;
+				}
+			}
+		}
+
+		private void AddPostExecuteTransformerForResultAggregate(MethodInfo aggregateMethodTemplate)
+		{
+			var elementType = Model.ResultTypeOverride;
+
+			LambdaExpression aggregateLambda;
+
+			if (elementType.IsNullableOrReference())
+			{
+				var inputListType = typeof(IQueryable<>).MakeGenericType(elementType);
```

Comment:
With my previous answer in mind, it would work with the `IEnumerable` version too for current usage of the post execute transformer.  I have made that choice for aligning with current input type of post execute transformers. True, there is an `AsQueryable` call just for supplying that input type to it. I really do not understand what were the design rationals behind the way this post execute transformer currently works. It looks to me as lot of overhead for no added value. (Enable  "merge" of execute transformer, but none of existing one seems mergeable to me. Maybe some serialization needs again. All that ends up in the query plan, which is serializable.)

With the current state of usage of this post execute transformer, it should works with `IEnumerable` methods too.